File: encode-and-decode-strings/flynn.cpp
/**
 * For the number of given strings N, and the length of the longest string M,
 * 
 * Encode
 * - Time complexity: O(N)
 * - Space complexity: O(1)
 * 
 * Decode
 * - Time complexity: O(NM)
 * - Space complexity: O(M)
 */

class Codec {
public:

    // Encodes a list of strings to a single string.
    string encode(vector<string>& strs) {
        string res = "";
        for (auto str : strs) {
            res += to_string(str.size());
            res.push_back('.');
            res += str;
        }
        return res;
    }

    // Decodes a single string to a list of strings.
    vector<string> decode(string s) {
        vector<string> res;

        auto it = s.begin();
        while (it != s.end()) {
            int str_size = 0;
            string tmp = "";

            while (*it != '.') {
                str_size = str_size * 10 + (*it - '0');
                it++;
            } 

            it++;

            for (int i = 0; i < str_size; i++) {
                tmp.push_back(*it);
                it++;
            }

            res.push_back(tmp);
        }

        return res;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.decode(codec.encode(strs));

File: serialize-and-deserialize-binary-tree/flynn.go
/*
풀이
- DFS를 이용하여 풀이합니다
Big O
- N: 노드의 수
- Serialize
  - Time complexity: O(N)
    - 모든 노드를 최대 1번 조회합니다
  - Space complexity: O(N)
    - buildString의 재귀 호출 스택의 깊이는 노드의 높이에 비례하여 증가하며, 노드의 높이는 최대 N입니다
    - 결과 string의 크기 또한 N에 비례하는 형태로 증가합니다
- Deserialize
  - Time complexity: O(N)
    - 모든 노드를 최대 1번 조회합니다
  - Space complexity: O(N)
    - data를 split한 배열의 크기가 N에 비례하여 증가합니다
	- buildTree의 재귀 호출 스택의 깊이는 노드의 높이에 비례하여 증가하며, 노드의 높이는 최대 N입니다
*/

import (
	"strconv"
	"strings"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const (
	DELIMITER = "|"
)

type Codec struct {
}

func Constructor() Codec {
	codec := Codec{}
	return codec
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	if root == nil {
		return ""
	}
	var sb strings.Builder
	buildString(&sb, root)
	return sb.String()
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	if data == "" {
		return nil
	}
	splitData := make([]string, 0, len(data)/2)
	splitData = strings.Split(data, DELIMITER)
	splitData = splitData[:len(splitData)-1]
	return buildTree(&splitData)
}

// ----- Helpers -----
func buildString(sb *strings.Builder, node *TreeNode) {
	if node == nil {
		sb.WriteString(DELIMITER)
		return
	}
	sb.WriteString(strconv.Itoa(node.Val))
	sb.WriteString(DELIMITER)
	buildString(sb, node.Left)
	buildString(sb, node.Right)
}

func buildTree(splitData *[]string) *TreeNode {
	val := (*splitData)[0]
	*splitData = (*splitData)[1:]
	if val == "" {
		return nil
	}
	node := &TreeNode{}
	intVal, _ := strconv.Atoi(val)
	node.Val = intVal
	node.Left = buildTree(splitData)
	node.Right = buildTree(splitData)
	return node
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */
